### ISSUE TITLE
Better Type error messages when FromPyObject fails on complex types

### ISSUE THREAD

**Issue by batconjurer** [PyO3/pyo3#1640]
I am currently working on a project where complex types are passed to Rust from Python and must be extracted to a Rust type. Most of the `FromPyObject` traits are implemented via the corresponding derive macro. 

However, the type errors that we get don't provide a lot of help in pinpointing what went wrong when incorrect input is given. Consider the following example 
```rust
// Rust
#[derive(FromPyObject)]
pub struct Inner {
   hi: i32,
   low: i32,
}

#[derive(FromPyObject)]
pub struct Outer {
    inner: Inner,
    extra: i32,
}

#[pyfunction]
fn foo(bar: Outer, baz: i32) {
    // --- snip -- //
}
```
```python
# Python 

class Inner(object):

    def __init__(self, hi, low):
        self.hi = hi
        self.low = low


class Outer(object):
    
    def __init__(self, inner, extra):
        self.inner = inner
        self.extra = extra
```
And then I call `foo(Outer(Inner(1.0, 0), 2), 3)` by accident. The error I get in Python will be something along the lines of
``` TypeError: 'float' object cannot be interpreted as an integer ```

What I would like to see is something along the lines of  
``` 
TypeError: Failed to extract Outer.inner because :: 
-- TypeErrror: Failed to extract Inner.hi because ::
-- TypeError: 'float' object cannot be interpreted as an integer 
```
although I am not at all attached to the formatting.

I am willing to work on this and playing around, I got the above to work by doing the following in 
https://github.com/PyO3/pyo3/tree/5446fe2062cb3bf11bf61bd4a2c58a7ed8b408d2/pyo3-macros-backend/src/from_pyobject.rs#L243
```rust
 let extractor = match &attrs.from_py_with {
     None => quote!(#get_field.extract().map_err(|inner| {
         let err_msg = format!("Failed to extract field {}.{} because:: \n-- {} ",
             stringify!(#self_ty),
             stringify!(#ident),
             inner);
             pyo3::exceptions::PyTypeError::new_err(err_msg)
     })?),
    Some(FromPyWithAttribute(expr_path)) => quote! (#expr_path(#get_field).
        map_err(|inner| {
            let err_msg = format!("Failed to extract field {}.{} because:: \n-- {} ",
                stringify!(#self_ty),
                stringify!(#ident),
                inner);
            pyo3::exceptions::PyTypeError::new_err(err_msg)
      })?),
};
```
Obviously this would need more work to support all ContainerTypes variants defined in that file (as well as enums), but I wanted some feedback on the idea and suggestions about the error message that is returned.

**Comment by davidhewitt** [comment]
👍 I am always in favour of better error messages! In terms of formatting, I quite like what `anyhow` does:

```
Error: Failed to read instrs from ./path/to/instrs.json

Caused by:
    No such file or directory (os error 2)
```

I guess in Python terms this might look more like this:

```
TypeError: failed to extract Outer.inner 

Caused by: 
    TypeError: failed to extract Inner.hi

Caused by:
    TypeError: 'float' object cannot be interpreted as an integer 
```

Our argument extraction code does this by colons, which could also work here:

```
TypeError: failed to extract Outer.inner: failed to extract Inner.hi: 'float' object cannot be interpreted as an integer 
```

----

Note that `FromPyObject` for enum variants might be a lot more complicated to format nicely. We try each enum variant in turn until one succeeds. A full error for enums might want to format the reason for each variant.

**Comment by batconjurer** [comment]
I quite like the second option you gave. The last one is hard for me to quickly read given that it is all in a single line. 

I think for struct variations, getting this implemented seems fairly straightforward. I haven't tried the enums yet, so I don't know yet how complicated that will become.

**Comment by mejrs** [comment]
+1. Getting mysterious errors from deep within libraries *sucks*.

What is more useful, I think, is a convenient way to customize the error.

After all, you might have users that don't know that they're using a module written in not-Python, in which case "failed to extract" might not be that useful to them - although mentioning `Inner.hi` helps.

Something like:
```rust
#[derive(FromPyObject)]
pub struct Inner {
   #[pyo3(conversion_error = "Inner.hi must be an integer")]
   hi: i32,
   low: i32,
}
```

Which could produce the same semantics and trace as reraising an error from another error in pure Python:
```
Traceback (most recent call last):
  File "user_file", line x, in <module>
    foo(Outer(Inner(1.0, 0), 2), 3)
TypeError: 'float' object cannot be interpreted as an integer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "..", line 4, in <module>
    raise TypeError("Inner.hi must be an integer") from e
TypeError: Inner.hi must be an integer
```

**Comment by batconjurer** [comment]
I think that the customizable message is a good idea as well.

As to your second point, do I understand that you proposed to mirror Python exceptions more closely in the formatting or are you just noting the nice parallel between how Python would propagate exceptions outwards and the current proposal?

**Comment by mejrs** [comment]
A bit of both, I think. The formatting doesn't really matter, but it would be nice if it was "python-esque".

**Comment by batconjurer** [comment]
@mejrs  One idea would be something like this (?):

```
Traceback (most recent call last):
  This error orginates in Rust
TypeError: 'float' object cannot be interpreted as an integer

The above exception was the direct cause of the following exception:
 
 Traceback (most recent call last):
     This error originates in Rust
 TypeError: Inner.hi must be an interger
```

This way it looks python-esque but also indicates clearly that the error is bubbling up from Rust code.

**Comment by batconjurer** [comment]
@davidhewitt For the enums, giving a reason for each variant might produce a very huge error message, especially if the the variants themselves contain complex types. While we could go that route, maybe it is an idea to have a brief error message and then by adding some sort of verbose attribute on the enum or a debug flag that produces fuller error messages.

**Comment by batconjurer** [comment]
So I seems trying to format the errors like 
```
Traceback (most recent call last):
  / ---- /
TypeError; Inner.hi must be an integer
```
Is going to difficult to pull off. This is because when we give an error message to `pyo3::exceptions::PyTypeError::new_err` it immediately tacks a `TypeError` to the beginning of the message. I think it will counter productive to modify this error type or do lots of string manipulation to get it into this format.

For now, I have take the approach suggested by @davidhewitt. However, I did come across something strange when running the tests.
```rust
#[derive(Debug, FromPyObject)]
pub enum Bar {
    #[pyo3(annotation = "str")]
    A(String),
    #[pyo3(annotation = "uint")]
    B(usize),
    #[pyo3(annotation = "int", transparent)]
    C(isize),
}

#[test]
fn test_err_rename() {
    let gil = Python::acquire_gil();
    let py = gil.python();
    let dict = PyDict::new(py);
    let f = Bar::extract(dict.as_ref());
   // -- snip -- //
}
```
produced the following error message
```
TypeError: Failed to extract type Bar

Caused by:
  TypeError: 'dict' object cannot be converted to 'Union[str,uint,int]'

TypeError: failed to extract inner field of Bar :: A

Caused by:
    TypeError: 'dict' object cannot be converted to 'PyString'

TypeError: failed to extract inner field of Bar :: B

Caused by:
    TypeError: 'dict' object cannot be interpreted as an integer

TypeError: failed to extract inner field of Bar :: C

Caused by:
    TypeError: 'dict' object cannot be interpreted as an integer
```
What is up with the `PyString` type? I would have assumed it would try to extract to a Rust string? Also, it is a bit unfortunate that the type names given in the attribute macros don't easily propagate to the types in the rest of the error message.

**Comment by davidhewitt** [comment]
Perhaps submit your current code as a PR and I can read it and comment? I'm sure we can make a number of improvments here.

(For example, I think that Python users would much rather see `str` instead of `PyString` or `String`. I have some ideas how to fix this but have not had the time to experiment!)

**Comment by batconjurer** [comment]
@davidhewitt I have made the PR, #1649.

**Comment by davidhewitt** [comment]
👍 thanks, I'm aiming give it a thorough review by the end of Saturday.

**Comment by davidhewitt** [comment]
This was implemented in #1649
